File: sourcerer-core/src/main/java/org/elder/sourcerer/EventRecord.java
package org.elder.sourcerer;

import com.google.common.collect.ImmutableMap;
import org.jetbrains.annotations.NotNull;

import java.time.Instant;
import java.util.UUID;

/**
 * Event event with metadata and related information, as read from an event repository.
 *
 * @param <T> The (base) type of events wrapped.
 */
public final class EventRecord<T> {
    @NotNull
    private final String streamId;
    private final int streamVersion;
    private final int aggregateVersion;
    @NotNull
    private final String eventType;
    @NotNull
    private final UUID eventId;
    @NotNull
    private final Instant timestamp;
    @NotNull
    private final ImmutableMap<String, String> metadata;
    @NotNull
    private final T event;

    public EventRecord(
            @NotNull final String streamId,
            final int streamVersion,
            final int aggregateVersion,
            @NotNull final String eventType,
            @NotNull final UUID eventId,
            @NotNull final Instant timestamp,
            @NotNull final ImmutableMap<String, String> metadata,
            final T event) {
        this.streamId = streamId;
        this.streamVersion = streamVersion;
        this.aggregateVersion = aggregateVersion;
        this.eventType = eventType;
        this.eventId = eventId;
        this.timestamp = timestamp;
        this.metadata = metadata;
        this.event = event;
    }

    @NotNull
    public String getStreamId() {
        return streamId;
    }

    /**
     * Gets the version (event sequence number) for the given event stream. This number relates to
     * the position of the events in the stream subscribed to or read from, which in the case of
     * projections / aggregate streams, may be different from the version relative to the stream
     */
    public int getStreamVersion() {
        return streamVersion;
    }

    /**
     * Gets the version (event sequence number) for the aggregate that the event relates to, i.e.
     * the original stream that the event was written to, as apposed to the event number on the
     * stream it was read from. In the case of aggregate streams / projections, this may differ
     * from the stream version.
     */
    public int getAggregateVersion() {
        return aggregateVersion;
    }

    @NotNull
    public String getEventType() {
        return eventType;
    }

    @NotNull
    public UUID getEventId() {
        return eventId;
    }

    @NotNull
    public Instant getTimestamp() {
        return timestamp;
    }

    @NotNull
    public ImmutableMap<String, String> getMetadata() {
        return metadata;
    }

    @NotNull
    public T getEvent() {
        return event;
    }

    @Override
    public String toString() {
        return "EventRecord{" +
                "streamId='" + streamId + '\'' +
                ", streamVersion=" + streamVersion +
                ", eventType='" + eventType + '\'' +
                ", eventId=" + eventId +
                ", metadata=" + metadata +
                ", event=" + event +
                '}';
    }
}

File: sourcerer-core/src/main/java/org/elder/sourcerer2/EventRecord.kt
package org.elder.sourcerer2

import com.google.common.collect.ImmutableMap
import java.time.Instant

/**
 * Event event with metadata and related information, as read from an event repository.
 *
 * @param <T> The (base) type of events wrapped.
 */
data class EventRecord<T>(
        /**
         * Gets a globally unique id for this event. This will be unique across all events and all
         * repositories.
         */
        val eventId: EventId,

        /**
         * The id of the stream that this event relates to. This id is unique only within a
         * particular repository, e.g. only across events describing the same type of aggregate.
         */
        val streamId: StreamId,

        /**
         * Gets the position of this event in the individual event stream that it was originally
         * written to (regardless of whether this came from reading that stream directly or from
         * subscribing to a repository as a whole.
         */
        val streamVersion: StreamVersion,

        /**
         * Gets the position of this event as it relates to others in the repository it was read
         * from, e.g. in relation to all other events representing aggregates of the same type.
         */
        val repositoryVersion: RepositoryVersion,

        /**
         * The application provided type of this particular event.
         */
        val eventType: String,

        /**
         * The time at which the storage engine recorded the event. This may be different from the
         * time of the actual business event that the event describes. It should not be relied on to
         * compare order of events, use stream or repository version instead.
         */
        val timestamp: Instant,

        /**
         * Other information about the event as provided when the event was persisted.
         */
        val metadata: ImmutableMap<String, String>,

        /**
         * The event data itself.
         */
        val event: T
)
